diff --git a/lib/cjs/Metadata/Class.js b/lib/cjs/Metadata/Class.js
index f6cc6fb510960bbc356e2b1cd4a64e2892c8a5fd..4376f55b7310c69a87ad23c8ff3d4f6aecacaea1 100644
--- a/lib/cjs/Metadata/Class.js
+++ b/lib/cjs/Metadata/Class.js
@@ -461,14 +461,14 @@ class ECClass extends SchemaItem_1.SchemaItem {
      */
     async *getAllBaseClasses() {
         for (const baseClassKey of this.schema.context.classHierarchy.getBaseClassKeys(this.key)) {
-            const baseClass = await this.schema.lookupItem(baseClassKey, ECClass);
+            const baseClass = await this.schema.context.getSchemaItem(baseClassKey, ECClass);
             if (baseClass)
                 yield baseClass;
         }
     }
     *getAllBaseClassesSync() {
         for (const baseClassKey of this.schema.context.classHierarchy.getBaseClassKeys(this.key)) {
-            const baseClass = this.schema.lookupItemSync(baseClassKey, ECClass);
+            const baseClass = this.schema.context.getSchemaItemSync(baseClassKey, ECClass);
             if (baseClass)
                 yield baseClass;
         }
diff --git a/lib/esm/Metadata/Class.js b/lib/esm/Metadata/Class.js
index 88355041a8540f211f0733bdcc8ab40177022bd4..3fbe050a3466c8723d5267b06d9ed26a7526f659 100644
--- a/lib/esm/Metadata/Class.js
+++ b/lib/esm/Metadata/Class.js
@@ -458,14 +458,14 @@ export class ECClass extends SchemaItem {
      */
     async *getAllBaseClasses() {
         for (const baseClassKey of this.schema.context.classHierarchy.getBaseClassKeys(this.key)) {
-            const baseClass = await this.schema.lookupItem(baseClassKey, ECClass);
+            const baseClass = await this.schema.context.getSchemaItem(baseClassKey, ECClass);
             if (baseClass)
                 yield baseClass;
         }
     }
     *getAllBaseClassesSync() {
         for (const baseClassKey of this.schema.context.classHierarchy.getBaseClassKeys(this.key)) {
-            const baseClass = this.schema.lookupItemSync(baseClassKey, ECClass);
+            const baseClass = this.schema.context.getSchemaItemSync(baseClassKey, ECClass);
             if (baseClass)
                 yield baseClass;
         }
